storage/index.go

```diff
@@ -21,61 +21,255 @@
+		i.metrics.InsertAfterClose.Inc(int64(len(inserts)))
+		return errDbIndexUnableToWriteClosed
+	}
+
+	var numErr int64
+	for _, insert := range inserts {
+		// FOLLOWUP(prateek): need to query before insert to ensure no duplicates && add test
+		err := i.active.segment.Insert(insert.doc)
+		if err != nil {
+			numErr++
+		} else {
+			insert.fns.OnIndexSuccess(i.active.expiryTime)
+		}
+		// NB: we need to release held resources so we un-conditionally execute the Finalize.
+		insert.fns.OnIndexFinalize()
+	}
+
+	if numErr != 0 {
+		i.metrics.AsyncInsertErrors.Inc(numErr)
+	}
+
+	return nil
+}
+
+func (i *nsIndex) Write(
 	id ident.ID,
-	tags ident.TagIterator,
+	tags ident.Tags,
+	fns onIndexSeries,
 ) error {
-	return errIndexingNotImplemented
+	d, err := i.doc(id, tags)
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		i.metrics.InsertAfterClose.Inc(1)
+		fns.OnIndexFinalize()
+		return errDbIndexUnableToWriteClosed
+	}
+
+	async := i.insertMode == index.InsertAsync
+	wg, err := i.insertQueue.Insert(d, fns)
+	i.RUnlock()
+
+	if err != nil {
+		fns.OnIndexFinalize()
+		return err
+	}
+
+	// once the write has been queued in the indexInsertQueue, it assumes
+	// responsibility for calling the lifecycle resource hooks.
+
+	if !async {
+		wg.Wait()
+	}
+
+	return nil
 }
 
-func (i *dbIndex) Query(
+func (i *nsIndex) Query(
 	ctx context.Context,
 	query index.Query,
 	opts index.QueryOptions,
 ) (index.QueryResults, error) {
-	return index.QueryResults{}, errIndexingNotImplemented
+	i.RLock()
+	if !i.isOpenWithRLock() {
+		i.RUnlock()
+		return index.QueryResults{}, errDbIndexUnableToQueryClosed
+	}
+
+	// FOLLOWUP(prateek): push down QueryOptions to restrict results
+	iter, err := i.active.segment.Query(query.Query)
+	i.RUnlock()
+	if err != nil {
+		return index.QueryResults{}, err
+	}
+
+	return index.QueryResults{
+		Iterator:   index.NewIterator(i.nsID, iter, i.opts),
+		Exhaustive: true,
+	}, nil
 }
 
-type databaseIndexWriter interface {
-	Write(
-		ctx context.Context,
-		namespace ident.ID,
-		id ident.ID,
-		tags ident.TagIterator,
-	) error
+func (i *nsIndex) isOpenWithRLock() bool {
+	return i.state == nsIndexStateOpen
+
 }
 
-type dbIndexNoOp struct{}
+func (i *nsIndex) Close() error {
+	i.Lock()
+	defer i.Unlock()
+	state := i.state
+	if state != nsIndexStateOpen {
+		return errDbIndexAlreadyClosed
+	}
 
-func (n dbIndexNoOp) Write(context.Context, ident.ID, ident.ID, ident.TagIterator) error {
-	return nil
+	i.state = nsIndexStateClosed
+	return i.insertQueue.Stop()
 }
 
-func (n dbIndexNoOp) Query(context.Context, index.Query, index.QueryOptions) (index.QueryResults, error) {
-	return index.QueryResults{}, nil
+func (i *nsIndex) doc(id ident.ID, tags ident.Tags) (doc.Document, error) {
+	fields := make([]doc.Field, 0, 1+len(tags))
+	fields = append(fields, doc.Field{
+		Name:      index.ReservedFieldNameID,
+		Value:     i.clone(id),
```

Comment:
This is the series ID? I think we can possibly avoid cloning this because it never gets finalized (WiredList and CommitLog and possibly some other things also depend on this behavior)

Reply:
@robskillington

Reply:
same as above:  I think you're right but I'd still prefer to explicitly copy. It makes the lifecycle a lot easier to reason about, and avoids issues if we ever do change to start finalizing ids.

Reply:
I guess if the index block size is different than m3db's blocksize that makes sense

Reply:
Yeah copying should be ok, it's only on insert thankfully.